docs/design/PT-166642114-miner-activated-consensus.md

```diff
@@ -0,0 +1,146 @@
+As miner signalling uses block header field `info`, the [current (i.e. old) consensus protocol version C1][fspecs] must be strictly greater than Minerva.
+
+#### Blocks
+
+##### Blocks connected to genesis
+
+The `aec_conductor` process, that is the entry point for all changes to the chain storage, enforces that each block being inserted is connected to genesis.
+By exploiting such current design, the expected consensus protocol version of a block can be computed whenever:
+- An untrusted block is meant to be able to be inserted in the storage.
+  This includes the cases of:
+  - Sync worker process intending to add one block to the storage.
+  - Gossip.
+  - User API.
+- A trusted block is meant to be able to be inserted in the storage.
+  These blocks are not validated.
+  This includes the cases of:
+  - Locally mined or signed block.
+- A block candidate (eventually leading to a trusted block) is prepared.
+  This includes the cases of key block and micro block.
+
+The check on such expected consensus protocol version
+whenever an untrusted block is meant to be able to be inserted in the storage - and any further deferred validation that that entails -
+can be performed outside of the context of the `aec_conductor` process.
+This follows the current design of the `aec_conductor` process,
+that relies on [most block validation having been performed in the caller's context](https://github.com/aeternity/aeternity/blob/d877a856648bd69cb1b473efa9c6149725d8d74c/apps/aecore/src/aec_conductor.erl#L1072-L1074):
+> Block validation is performed in the caller's context for
+> external (gossip/sync) blocks and we trust the ones we
+> produce ourselves.
+
+(Blocks from the internal user API are considered equivalent to blocks received via gossip, hence untrusted and validated.)
+
+Keeping these checkes outside of the context of the `aec_conductor` process is relevant
+in order to prevent passing to the `aec_conductor` process too many invalid blocks
+(either for chain forks caused by the change in consensus protocol version, or for malicious activity).
+Specifically, the contexts in which the consensus version must be checked beforehand outside of the context of the `aec_conductor` process are:
+- The sync worker process.
+- The peer connection / gossip process.
+- The HTTP user API worker.
+
+For blocks mined or signed locally, the conductor must instruct the generation of the block candidate with the correct consensus protocol version.
+
+###### Asynchronous computation of expected consensus protocol version
+
+In order to prevent long and potentially duplicated computational efforts - requiring multiple lookups into the chain storage -
+the computation of the outcome of the miner signalling should be asynchronous and lazy.
+
+Whenever a block at height preceding the signalling block interval end is inserter in the chain storage,
+the `aec_conductor` spawns - and keeps track of - a worker responsible for
```

Comment:
As discussed in the chat. A worker can be spawned at height `HE - 1` (signalling end interval - 1) and not necessarily for each inserted block within the signalling interval.

Reply:
Also, the worker can be spawned at hard fork height - 1, so we know the result at fork height. As the signalling computation result could take a significant amount of time, it must be done earlier than hard fork height - 1.

Reply:
> As discussed in the chat. A worker can be spawned at height `HE - 1` (signalling end interval - 1) and not necessarily for each inserted block within the signalling interval.

Clarified in new commit baae00f12

Reply:
> Also, the worker can be spawned at hard fork height - 1, so we know the result at fork height. As the signalling computation result could take a significant amount of time, it must be done earlier than hard fork height - 1.

I believe this is addressed, as the worker is documented as spawned at "HE - 1".